pkg/db/db_common/wait_connection.go

```diff
@@ -35,27 +39,37 @@ func WaitForPool(ctx context.Context, db *pgxpool.Pool) (err error) {
 	}
 }
 
-// WaitForConnection waits for the db to start accepting connections and returns true
-// returns false if the dbClient does not start within a stipulated time,
-func WaitForConnection(ctx context.Context, db *pgx.Conn) (err error) {
+// WaitForConnection PINGs the DB - retrying after a backoff of constants.ServicePingInterval - but only for constants.DBConnectionTimeout
+// returns the error from the database if the dbClient does not respond with after a timeout
+func WaitForConnection(ctx context.Context, connection *pgx.Conn) (err error) {
 	utils.LogTime("db.waitForConnection start")
 	defer utils.LogTime("db.waitForConnection end")
 
-	pingTimer := time.NewTicker(constants.ServicePingInterval)
-	timeoutAt := time.After(constants.DBConnectionTimeout)
-	defer pingTimer.Stop()
+	timeoutCtx, cancel := context.WithTimeout(ctx, constants.DBConnectionTimeout)
+	defer func() {
+		cancel()
+	}()
 
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-pingTimer.C:
-			err = db.Ping(ctx)
-			if err == nil {
-				return
+	retryBackoff := retry.WithMaxDuration(
+		constants.DBConnectionTimeout,
+		retry.NewConstant(constants.ServicePingInterval),
+	)
+
+	retryErr := retry.Do(ctx, retryBackoff, func(ctx context.Context) error {
+		pingErr := connection.Ping(timeoutCtx)
+		if pingErr != nil {
+			if pgErr, ok := pingErr.(*pgconn.PgError); ok && pgErr.SQLState() == "57P03" {
```

Comment:
I haven't checked this but we should also wait if Postgres is doing auto recovery.

In a large database this can take "a long time".